### PR TITLE
Replace MPSGraph count_nonzero with custom Metal kernel

### PR DESCRIPTION
**Pull request by malfet**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #180725
* #180709

Migrate count_nonzero to native Metal by extending the existing sum reduction kernels with a third LoadMode variant that counts nonzero elements.

Refactor the NAN_TO_ZERO bool template parameter in ReduceOps.metal to a LoadMode enum with three values: `LOAD_IDENTITY` (sum), `LOAD_NAN_TO_ZERO` (nansum), and `LOAD_NONZERO` (count_nonzero — contributes 1 per nonzero, 0 otherwise). The sum_reduction, sum_reduction_outer, and sum_reduction_inner templates all share the same load dispatch; the compiler eliminates dead branches per instantiation. Even though output is 64-bit integer,  accumulates partial counts in `uint32` rather than `long` as tensor dimension is always less that in32 on Metal.

Register count_nonzero_{,_outer,_inner}_<TI>_long kernels for all supported input dtypes (float/half/bfloat/{int,long,short,char,uchar}/bool/float2/half2). Host-side count_nonzero_mps now builds a TensorIterator via make_reduction and dispatches to sum_nansum_kernel_mps with prefix "count_nonzero_"

Remove `MPSReductionType::COUNT_NONZERO` and its MPSGraph branch.

Tweaked the two-pass full-reduction path to always dispatch pass 2 with `LOAD_IDENTITY` (plain sum), regardless of pass-1 mode: pass 1 has already applied the load-mode transform (NaN-to-zero, nonzero-as-one), so pass 2 only needs to combine the partials.

Benchmark vs torch 2.11 MPSGraph (median us, M4 Max):

| shape       | op    | dtype   | 2.11 | new  | speedup |
|-------------|-------|---------|------|------|---------|
| (10000,)    | full  | float32 | 14.0 |  8.1 | 1.73x   |
| (1000000,)  | full  | float32 | 14.2 | 10.8 | 1.31x   |
| (1024,1024) | full  | float32 | 16.4 | 11.6 | 1.41x   |
| (1024,1024) | dim=0 | float32 | 13.0 |  6.0 | 2.18x   |
| (1024,1024) | dim=1 | float32 | 12.9 |  4.9 | 2.62x   |
| (4096,4096) | full  | float32 |131.2 |121.9 | 1.08x   |
| (4096,4096) | dim=0 | float32 |132.4 |127.1 | 1.04x   |
| (1024,1024) | full  | int64   | 15.5 | 12.8 | 1.21x   |
| (1024,1024) | dim=0 | int64   | 12.9 |  7.6 | 1.70x   |
| (1024,1024) | dim=0 | float16 | 14.7 |  4.7 | 3.13x   |
| (1024,1024) | full  | bool    | 15.1 | 11.2 | 1.34x   |
| (1024,1024) | dim=0 | bool    | 13.2 |  4.8 | 2.76x   |
| (4096,4096) | dim=0 | bool    | 48.8 | 26.2 | 1.86x   |

Dim reductions see parity or 1.7–3x speedups across dtypes.

<details>
<summary>Benchmark script</summary>

```python
import torch
import torch.utils.benchmark as benchmark
import itertools

shapes_and_dims = [
    ((10_000,), None),
    ((1_000_000,), None),
    ((1024, 1024), None),
    ((1024, 1024), 0),
    ((1024, 1024), 1),
    ((4096, 4096), None),
    ((4096, 4096), 0),
    ((4096, 4096), 1),
]
sparsities = [0.2, 0.5, 0.8]

def make_input(shape, sparsity, dtype, device):
    if dtype == torch.bool:
        return torch.rand(shape, device=device) < sparsity
    x = torch.randn(shape, device=device, dtype=dtype)
    mask = torch.rand(shape, device=device) >= sparsity
    x[mask] = 0
    return x

def bench(x, dim):
    stmt = "torch.count_nonzero(x)" if dim is None else "torch.count_nonzero(x, dim=d)"
    t = benchmark.Timer(stmt=stmt, globals={"torch": torch, "x": x, "d": dim}, num_threads=1)
    for _ in range(2):
        result = t.timeit(500)
    return result.median

for (shape, dim), sparsity in itertools.product(shapes_and_dims, sparsities):
    x = make_input(shape, sparsity, torch.float32, "mps")
    m = bench(x, dim) * 1e6
    print(f"{str(shape):>14} {str(dim):>5} {sparsity:>5.2f} {m:>12.2f}")
```
</details>

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

**Comment by hvaara** [comment]
This PR partially fixes https://github.com/pytorch/pytorch/issues/178497
